### PR TITLE
libpulsar: init at 2.9.1

### PR DESCRIPTION
**Pull request by CorbanR**
###### Motivation for this change
Adding corbanr user to maintainers/maintainer-list.nix

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bobby285271** [comment]
Can you also add yourself to the package you wish to maintain in this PR?

**Comment by CorbanR** [comment]
For sure! I was going to do them in separate PRs, but I'm happy to make it one.

**Comment by 0x4A6F** [comment]
You might want to carve out the changes to `maintainers/maintainer-list.nix` into the first commit and the changes to `pkgs` into the commit `libpulsar: init at 2.9.1` according to [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

**Comment by CorbanR** [comment]
@0x4A6F thanks for the advice, will do. Will also re-read the contrib guide.

**Comment by CorbanR** [comment]
@bobby285271 @0x4A6F how do I request reviews? I don't have perms, and Im not seeing documentation on it.

**Review by 0x4A6F — CHANGES_REQUESTED** [review]
Thanks for your changes. Found just a typo, otherwise it looks good to go from my standpoint. Someone with more insight could have a second look.
You might request reviewers after this pr is merged and you'll be invited to [nixpkgs-maintainers](https://github.com/orgs/NixOS/teams/nixpkgs-maintainers).

Do you have any consumers for this lib?

File: pkgs/development/libraries/libpulsar/default.nix [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+, zstd ? null
+}:
+
+let
+  /*
+    Check if null or false
+    Example:
+    let result = enableFeature null
+    => "OFF"
+    let resule = enableFeature false
```
[the remark this answers is not in the document]
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: Thank you for the review! Fixed!. Me and a few people I work with use this lib in order for https://github.com/instructure/pulsar-client-ruby to work.

**Review by 0x4A6F — APPROVED** [review]

**Comment by CorbanR** [comment]
@0x4A6F do you know if will someone with merge permissions eventually get to this? I know there are a ton of pull requests, just curious if you have any experience with turn around time?

File: pkgs/development/libraries/libpulsar/default.nix [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  # If boost has python enabled, then boost-python package will be installed which is used by libpulsars python wrapper
+  boost = if enablePython then boost17x.override { inherit stdenv enablePython; python = python3; } else boost17x;
+in
+stdenv.mkDerivation rec {
+  pname = "libpulsar";
+  version = "2.9.1";
+
+  src = fetchurl {
+    sha512 = "sha512-NKHiL7D/Lmnn6ICpQyUmmQYQETz4nZPJU9/4LMRDUQ3Pck6qDh+t6CRk+b9UQ2Vb0jvPIGTjEsSp2nC7TJk3ug==";
+    url = "https://archive.apache.org/dist/pulsar/pulsar-${version}/apache-pulsar-${version}-src.tar.gz";
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:48]: Some other Apache packages use `mirror://apache/` - might want to do the same here?

File: pkgs/development/libraries/libpulsar/default.nix [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  installCheckPhase = ''
+    echo ${lib.escapeShellArg ''
+      #include <pulsar/Client.h>
+      int main (int argc, char **argv) {
+        pulsar::Client client("pulsar://localhost:6650");
+        return 0;
+      }
+    ''} > test.cc
+        ${if stdenv.isDarwin
+        then "$CC++ test.cc -L $out/lib -I $out/include -lpulsar -o test"
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:86]: Doesn't `$CXX` point to the C++ compiler in both cases? Would get rid of the `if` construct

File: pkgs/development/libraries/libpulsar/default.nix [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+, log4cxx ? null
+, openssl
+, pkg-config
+, protobuf
+, python3 ? null
+, snappy ? null
+, stdenv
+, zlib ? null
+, zstd ? null
+}:
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:22]: From nixpkgs [conventions](https://nixos.org/manual/nixpkgs/stable/#chap-conventions)

_Arguments should be listed in the order they are used, with the exception of lib, which always goes first._

File: pkgs/development/libraries/libpulsar/default.nix [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    sha512 = "sha512-NKHiL7D/Lmnn6ICpQyUmmQYQETz4nZPJU9/4LMRDUQ3Pck6qDh+t6CRk+b9UQ2Vb0jvPIGTjEsSp2nC7TJk3ug==";
+    url = "https://archive.apache.org/dist/pulsar/pulsar-${version}/apache-pulsar-${version}-src.tar.gz";
+  };
+
+  sourceRoot = "apache-pulsar-${version}-src/pulsar-client-cpp";
+
+  # clang-tools needed for clang-format etc
+  # filter out null values so users can override optional values(such as gtest)
+  nativeBuildInputs = lib.filter (x: x != null) ([ cmake boost protobuf python3 pkg-config gtest ]
+    ++ lib.optional stdenv.isDarwin [ clang clangTools ]
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:56]: Doesn't `lib.optional` return a list of lists like this?
→ `lib.optionals`

Same for other invocations.

File: pkgs/development/libraries/libpulsar/default.nix [hunk 6]
```diff
@@ -0,0 +1,98 @@
+{ boost17x
+, clang
+, clang-tools
+, cmake
+, curl
+, enablePython ? false
+, fetchurl
+, gcc
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:8]: Does that make sense on darwin?

Also I guess it's part of stdenv, so doesn't need to be listed explicitly.

File: pkgs/development/libraries/libpulsar/default.nix [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  # clang-tools needed for clang-format etc
+  # filter out null values so users can override optional values(such as gtest)
+  nativeBuildInputs = lib.filter (x: x != null) ([ cmake boost protobuf python3 pkg-config gtest ]
+    ++ lib.optional stdenv.isDarwin [ clang clangTools ]
+    ++ lib.optional stdenv.isLinux [ gcc clangTools ]);
+
+  # Filter out null vaules so users can override optional values(such as log4cxx)
+  buildInputs = lib.filter (x: x != null) ([ boost jsoncpp log4cxx openssl protobuf snappy zstd curl zlib ] ++ lib.optional enablePython [ python3 ]);
+
+  # since we cant expand $out in cmakeFlags
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:62]: I think you might be able to use `${placeholder "out"}`

File: pkgs/development/libraries/libpulsar/default.nix [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  sourceRoot = "apache-pulsar-${version}-src/pulsar-client-cpp";
+
+  # clang-tools needed for clang-format etc
+  # filter out null values so users can override optional values(such as gtest)
+  nativeBuildInputs = lib.filter (x: x != null) ([ cmake boost protobuf python3 pkg-config gtest ]
+    ++ lib.optional stdenv.isDarwin [ clang clangTools ]
+    ++ lib.optional stdenv.isLinux [ gcc clangTools ]);
+
+  # Filter out null vaules so users can override optional values(such as log4cxx)
+  buildInputs = lib.filter (x: x != null) ([ boost jsoncpp log4cxx openssl protobuf snappy zstd curl zlib ] ++ lib.optional enablePython [ python3 ]);
```
Comment (ius) [inline pkgs/development/libraries/libpulsar/default.nix:60]: Consider listing one list element per line for legibility.

That way it also doesn't wrap on GitHub's diff view ;)

**Review by ius — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/libpulsar/default.nix (hunk 2 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix:48]: Good call!

File: pkgs/development/libraries/libpulsar/default.nix (hunk 3 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: Yes. Thank you for all your comments, I have made multiple updates, removing some cruft!

File: pkgs/development/libraries/libpulsar/default.nix (hunk 4 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix:25]: I think the arguments are more closely ordered to how they are used now. Thank you!

File: pkgs/development/libraries/libpulsar/default.nix (hunk 5 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: I got rid of this function and added some arguments such as `snappySupport`. Then used `lib.optional`. I also got rid of the `null` defaults as they were superfluous.

File: pkgs/development/libraries/libpulsar/default.nix (hunk 7 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: I was able to get rid of the variable entirely now.

File: pkgs/development/libraries/libpulsar/default.nix (hunk 6 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: I have removed both `gcc`, and `clang` from the argument list. You are right, they don't need to be listed explicitly.

File: pkgs/development/libraries/libpulsar/default.nix (hunk 8 above)
Comment (CorbanR) [reply pkgs/development/libraries/libpulsar/default.nix]: I added a `let` with dependencies that are in both `buildInputs` and `nativeBuildInputs`

**Review by 0x4A6F — CHANGES_REQUESTED** [review]
Check correct usage of
```
 ++ lib.optionals condition [ listItem1 listItem2 ]
```
vs
```
 ++ lib.optional condition singleItem;
```

**Comment by CorbanR** [comment]
> Check correct usage of
> 
> ```
>  ++ lib.optionals condition [ listItem1 listItem2 ]
> ```
> 
> vs
> 
> ```
>  ++ lib.optional condition singleItem;
> ```

Thank you!, I fixed the use of `lib.optional`, i'm actually surprised it worked in my local testing the otherway!

**Review by 0x4A6F — APPROVED** [review]
LGTM

Thanks, ping me in a week if no one wants to merge.

**Pull request merged**